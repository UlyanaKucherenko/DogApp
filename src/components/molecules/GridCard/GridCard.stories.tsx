import type { Meta } from '@storybook/react';

import { Box } from '@mui/material';
import { GridCard, GridCardProps } from '.';

const meta: Meta<GridCardProps> = {
  title: 'Molecules/GridCard',
  component: GridCard
};
export default meta;

export const Main = () => (
  <Box sx={{ width: '500px', height: '400px' }}>
    <GridCard
      imgUrl="http://via.placeholder.com/640x360"
      id="1"
      isFavorite={false}
      cardType="favorite"
    />
  </Box>
);

export const Upload = () => (
  <Box sx={{ width: '500px', height: '400px' }}>
    <GridCard
      imgUrl="http://via.placeholder.com/640x360"
      id="1"
      isFavorite={false}
      cardType="upload"
    />
  </Box>
);
