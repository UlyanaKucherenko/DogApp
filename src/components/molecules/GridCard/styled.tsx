import { Card } from '@mui/material';
import { styled } from '@mui/material/styles';
import BorderedBox from '../../atoms/BorderedBox';

const CardBorder = styled(BorderedBox)(() => ({
  height: '100%'
}));

const CardBox = styled(Card)(() => ({
  borderRadius: '20px',
  boxShadow: 'none',
  position: 'relative',
  height: '100%'
}));

export { CardBorder, CardBox };
