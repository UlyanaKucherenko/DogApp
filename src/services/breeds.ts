import { api } from './api';

export const breedsApi = api.injectEndpoints({
  endpoints: build => ({
    getBreeds: build.query<Breed[], void>({
      query: () => 'breeds'
    }),
    getBreedById: build.query<BreedDetails, string>({
      query: breedId => `breeds/${breedId}`
    })
  })
});

export const { useGetBreedsQuery, useGetBreedByIdQuery } = breedsApi;

export interface Breed {
  id: string;
  name: string;
  temperament: string;
}

export interface BreedDetails extends Breed {
  weight: {
    metric: string;
  };
  height: {
    metric: string;
  };
  lifeSpan: string;
  origin: string;
}
