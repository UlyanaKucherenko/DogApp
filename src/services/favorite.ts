import { api } from './api';

export interface Favorite {
  image_id: string;
  sub_id?: string;
}

export type FavoritesResponse = {
  id: number;
  user_id: string;
  image_id: string;
  sub_id: string;
  created_at: string;
  image: {
    id: string;
    url: string;
  };
};

export const favoritesApi = api.injectEndpoints({
  endpoints: build => ({
    addFavorite: build.mutation<unknown, Favorite>({
      query: body => ({
        url: `favourites`,
        method: 'POST',
        body
      }),
      invalidatesTags: [{ type: 'Favorites', id: 'LIST' }]
    }),

    deleteFavorite: build.mutation<unknown, { favourite_id: number }>({
      query: ({ favourite_id }) => ({
        url: `favourites/${favourite_id}`,
        method: 'DELETE'
      }),
      invalidatesTags: [{ type: 'Favorites', id: 'LIST' }]
    }),

    getFavorites: build.query<FavoritesResponse[], void>({
      query: () => ({
        url: `favourites`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Favorites', id }) as const),
        { type: 'Favorites' as const, id: 'LIST' }
      ]
    })
  })
});

export const {
  useAddFavoriteMutation,
  useDeleteFavoriteMutation,
  useGetFavoritesQuery
} = favoritesApi;
