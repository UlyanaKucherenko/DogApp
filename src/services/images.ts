import { api } from './api';

export interface Images {
  id: string;
  url: string;
  width: number;
  height: number;
  mime_type: string;
  breeds: [
    {
      id: number;
      name: string;
      weight: string;
      height: string;
      life_span: string;
      reference_image_id: string;
    }
  ];
  categories: [];
}

export interface UploadImage {
  breeds: [];
  id: string;
  url: string;
  width: number;
  height: number;
  sub_id: string;
  created_at: string;
  original_filename: string;
  breed_ids: string;
}

export type ImagesUploadResponse = {
  id: string;
  url: string;
  width: number;
  height: number;
  original_filename: string;
  pending: number;
  approved: number;
};

export interface BreedImage {
  id: string;
  url: string;
  reference_image_id: string;
}

type ImagesResponse = {
  [breedId: string]: BreedImage[];
};

export const imagesApi = api.injectEndpoints({
  endpoints: build => ({
    getBreedImages: build.query<ImagesResponse, string>({
      query: imageId => ({
        url: `images/${imageId}`
      }),
      providesTags: id => [{ type: 'Images', id }]
    }),

    getImagesId: build.query<
      ImagesResponse,
      {
        limit?: number;
        page?: number;
        mime_type?: string;
        order?: string;
        breed_id?: number;
      }
    >({
      query: ({
        limit = 10,
        page = 0,
        mime_type = 'all',
        breed_id,
        order = 'RANDOM'
      }) => ({
        url: `images/search?limit=${limit}&page=${page}&mime_types=${mime_type}&order=${order}&breed_id=${breed_id}`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Images', id }) as const),
        { type: 'Images' as const, id: 'LIST' }
      ]
    }),

    getImages: build.query<
      ImagesResponse,
      {
        limit?: number;
        page?: number;
        mime_type?: string;
        order?: string;
        breed_id?: number;
      }
    >({
      query: ({
        limit = 10,
        page = 0,
        mime_type = 'all',
        breed_id,
        order = 'RANDOM'
      }) => ({
        url: `images/search?limit=${limit}&page=${page}&mime_types=${mime_type}&order=${order}`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Images', id }) as const),
        { type: 'Images' as const, id: 'LIST' }
      ]
    }),

    uploadImage: build.mutation<Images, FormData>({
      query(formData) {
        return {
          url: `images/upload`,
          method: 'POST',
          body: formData
        };
      },
      invalidatesTags: [{ type: 'Images', id: 'LIST' }]
    }),
    uploadImagesList: build.query<
      UploadImage[],
      {
        limit?: number;
        page?: number;
      }
    >({
      query: ({ limit = 10, page = 0 }) => ({
        url: `images/?limit=${limit}&page=${page}`
      }),
      providesTags: (result = []) => [
        ...result.map(({ id }) => ({ type: 'Images', id }) as const),
        { type: 'Images' as const, id: 'LIST' }
      ]
    })
  })
});

export const {
  useGetBreedImagesQuery,
  useGetImagesIdQuery,
  useGetImagesQuery,
  useUploadImageMutation,
  useUploadImagesListQuery
} = imagesApi;
