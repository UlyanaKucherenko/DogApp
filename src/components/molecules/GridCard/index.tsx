import { CardMedia, alpha } from '@mui/material';
import theme from '../../../theme/theme';
import { CardBorder, CardBox } from './styled';
import { GridCardActions } from '../GridCardActions';

export interface GridCardProps {
  id: number | string;
  imgUrl: string;
  isFavorite?: boolean;
  cardType?: 'favorite' | 'upload';
}

export const GridCard = ({
  id,
  imgUrl,
  isFavorite = false,
  cardType
}: GridCardProps) => (
  <CardBorder showHovered boxShadow={alpha(theme.palette.primary.main, 1)}>
    <CardBox>
      <CardMedia
        component="img"
        image={imgUrl || 'http://via.placeholder.com/640x360'}
        alt="item"
        sx={{ height: '100%', objectFit: 'cover' }}
      />
      <GridCardActions id={id} isFavorite={isFavorite} cardType={cardType} />
    </CardBox>
  </CardBorder>
);
