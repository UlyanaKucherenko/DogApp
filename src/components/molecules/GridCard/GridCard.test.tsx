import { composeStories } from '@storybook/react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import * as stories from './GridCard.stories';
import { store } from '../../../services/store';

describe('GridCard', () => {
  const { Main } = composeStories(stories);
  test('should match snapshot Main', () => {
    const { container } = render(
      <Provider store={store}>
        <Main />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
