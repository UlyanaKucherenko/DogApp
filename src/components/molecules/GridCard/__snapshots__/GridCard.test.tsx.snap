// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GridCard should match snapshot Main 1`] = `
.emotion-0 {
  width: 500px;
  height: 400px;
}

.emotion-1 {
  box-shadow: rgba(146, 31, 237, 1);
  border-radius: 20px;
  box-shadow: 6px 6px 0px 0px rgba(0, 0, 0, 1);
  -webkit-transition: box-shadow 0.3s ease-in-out;
  transition: box-shadow 0.3s ease-in-out;
  height: 100%;
}

.emotion-1:hover {
  box-shadow: 6px 6px 8px 0px rgba(25, 118, 210, 1);
}

.emotion-1:active {
  box-shadow: 6px 6px 8px 0px rgba(25, 118, 210, 1);
}

.emotion-2 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
  overflow: hidden;
  border-radius: 20px;
  box-shadow: none;
  position: relative;
  height: 100%;
}

.emotion-3 {
  display: block;
  -webkit-background-size: cover;
  background-size: cover;
  background-repeat: no-repeat;
  -webkit-background-position: center;
  background-position: center;
  width: 100%;
  object-fit: cover;
  height: 100%;
  object-fit: cover;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  padding: 24px 24px 20px 24px;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  width: 100%;
  height: 100%;
  background-color: transparent;
  -webkit-transition: background-color 0.3s ease-in-out;
  transition: background-color 0.3s ease-in-out;
}

.emotion-4>:not(:first-of-type) {
  margin-left: 8px;
}

.emotion-4:hover {
  background-color: rgba(0, 0, 0, 0.45);
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  width: 100%;
}

.emotion-6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-6::-moz-focus-inner {
  border-style: none;
}

.emotion-6.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-6 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-6:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-6:hover {
    background-color: transparent;
  }
}

.emotion-6.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-7 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<div>
  <div
    class="MuiBox-root emotion-0"
  >
    <div
      class="MuiBox-root emotion-1"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root emotion-2"
      >
        <img
          alt="item"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img emotion-3"
          src="http://via.placeholder.com/640x360"
        />
        <div
          class="MuiCardActions-root MuiCardActions-spacing emotion-4"
        >
          <div
            class="MuiBox-root emotion-5"
          >
            <button
              aria-label="add to favorites"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-6"
              tabindex="0"
              type="button"
            >
              <test-file-stub
                aria-hidden="true"
                classname="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                focusable="false"
              />
              <span
                class="MuiTouchRipple-root emotion-7"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
